src/shutdown.rs

```diff
@@ -0,0 +1,96 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use std::sync::Arc;
+use tokio::signal;
+use tokio_util::sync::CancellationToken;
+use tracing::{error, info};
+
+#[async_trait]
+pub trait ShutdownHook: Send + Sync {
+  async fn shutdown(&self) -> Result<()>;
+}
+
+pub struct ShutdownCoordinator {
+  token: CancellationToken,
+  hooks: Vec<Arc<dyn ShutdownHook>>,
+}
+
+impl ShutdownCoordinator {
+  pub fn new() -> Self {
+    Self {
+      token: CancellationToken::new(),
+      hooks: Vec::new(),
+    }
+  }
+
+  pub fn token(&self) -> CancellationToken {
+    self.token.clone()
+  }
+
+  pub fn register_hook(&mut self, hook: Arc<dyn ShutdownHook>) {
+    self.hooks.push(hook);
+  }
```

Comment:
I missed where this is used, but seems relevant for how actors will register themselves 🤔